smithy-model/src/main/java/software/amazon/smithy/model/shapes/MapShape.java

```diff
@@ -43,7 +44,19 @@ public final class MapShape extends Shape implements ToSmithyBuilder<MapShape> {
     private MapShape(Builder builder) {
         super(builder, false);
         key = getRequiredMixinMember(builder, builder.key, "key");
+        ShapeId expectedKeyId = getId().withMember("key");
+        if (!key.getId().equals(expectedKeyId)) {
+            throw new SourceException(String.format(
+                    "Expected member of `%s` to have an ID of `%s` but found `%s`",
```

Comment:
Would it be possible to move this validation here and below into the builder, like what you did with CollectionShape?

Reply:
I confused myself with this one. I did add these checks in the MapShape constructor, not the one in CollectionShape. They aren't necessary though because the constructor calls [validateMemberShapeIds](https://github.com/awslabs/smithy/blob/5636d15d067a63613cb9a1a069585fc81312df6b/smithy-model/src/main/java/software/amazon/smithy/model/shapes/Shape.java#L165) which does the same thing. I'll push an update to remove them.